### PR TITLE
Require version 3,11 of Cabal to support --semaphore flag

### LINKED ISSUE #9197: --semaphore is not compatible with build-type: Custom
**Describe the bug**
Cabal HEAD fails to build any package with `build-type: Custom` when `-j --semaphore` is passed.

**To Reproduce**

```
$ cabal unpack ghc-paths-0.1.0.12 && cd ghc-paths-0.1.0.12 && cabal build -w ghc-9.8.0.20230809 -j --semaphore --allow-newer
Resolving dependencies...
Build profile: -w ghc-9.8.0.20230809 -O1
In order, the following will be built (use -v for more details):
 - ghc-paths-0.1.0.12 (lib:ghc-paths) (first run)
Created semaphore called cabal_semaphore_3 with 8 slots.
[1 of 2] Compiling Main             ( /private/tmp/ghc-paths-0.1.0.12/dist-newstyle/build/aarch64-osx/ghc-9.8.0.20230809/ghc-paths-0.1.0.12/setup/setup.hs, /private/tmp/ghc-paths-0.1.0.12/dist-newstyle/build/aarch64-osx/ghc-9.8.0.20230809/ghc-paths-0.1.0.12/setup/Main.o )
[2 of 2] Linking /private/tmp/ghc-paths-0.1.0.12/dist-newstyle/build/aarch64-osx/ghc-9.8.0.20230809/ghc-paths-0.1.0.12/setup/setup
Configuring ghc-paths-0.1.0.12...
unrecognized 'build' option `--semaphore=cabal_semaphore_3'
Error: cabal: Failed to build ghc-paths-0.1.0.12.
```

**Expected behavior**
The build should complete successfully. If I remove `-j --semaphore`, all works well.

**System information**
 - Operating system: macOS
 - `cabal`: HEAD (41225bb43)
 - `ghc`: 9.8.0.20230809

### PR DESCRIPTION
**Pull request by mpickering**
Fixes #9197

### CODE REVIEW

File: cabal-install/src/Distribution/Client/ProjectPlanning.hs [hunk 1]
```diff
@@ -4294,7 +4294,7 @@ setupHsBuildFlags par_strat elab _ verbosity builddir =
     , buildDistPref = toFlag builddir
     , buildNumJobs = mempty -- TODO: [nice to have] sometimes want to use toFlag (Just numBuildJobs),
     , buildUseSemaphore =
-        if elabSetupScriptCliVersion elab >= mkVersion [3, 9, 0, 0]
+        if elabSetupScriptCliVersion elab >= mkVersion [3, 11, 0, 0]
```
Comment (sheaf) [inline cabal-install/src/Distribution/Client/ProjectPlanning.hs:4297]:
```suggestion
        if elabSetupScriptCliVersion elab >= mkVersion [3, 11, 0, 0]
          -- Cabal 3.11 is the first version that supports parallelism semaphores
```

**Review by Mikolaj — APPROVED** [review]
Thank you.

**Review by Kleidukos — APPROVED** [review]

**Comment by Mikolaj** [comment]
Ideally after adding the comment @sheaf proposes, I guess?

**Comment by andreabedini** [comment]
I just went head, committed @sheaf's suggestion and updated the branch.

File: cabal-install/src/Distribution/Client/ProjectPlanning.hs [hunk 2]
```diff
@@ -4283,7 +4283,8 @@ setupHsBuildFlags par_strat elab _ verbosity builddir =
...
+          -- Cabal 3.11 is the first version that supports parallelism semaphores
           then par_strat
```
Comment (alt-romes) [inline cabal-install/src/Distribution/Client/ProjectPlanning.hs:4288]:
```suggestion
          then -- Cabal 3.11 is the first version that supports parallelism semaphores
            par_strat
```

Comment (alt-romes) [reply cabal-install/src/Distribution/Client/ProjectPlanning.hs:4288]: CI fails because of formoulu, this should fix it.

Comment (Mikolaj) [reply cabal-install/src/Distribution/Client/ProjectPlanning.hs:4288]: @mpickering care to set the "merge_me` label, after committing this change, if it looks right to you?

**Pull request merged**